### PR TITLE
Dust api client bundle types

### PR DESCRIPTION
**Pull request by Fraggle**
## Description

On the path to package DustApi client, the intent is to have the dust-client package declare all schemas and types used by public (/v1/...) endpoints and then only use them in the /v1/ files so we have a clean separation on what the public API is supposed to return vs the private API and rely on typescript to catch any change that might break the public API contract.

To do that, we:
- Removed the dependancy of DustApi client on dust-types
- Recreated all types in `zod` format in `sdks/js/types.ts`
- Exported request _schemas_ and response _types_ and imported them in all v1 endpoints.
- Refactored when necessary in some cases where v1 endpoint was using private api stuff directly (moved the logic function in /lib/ and separated the public and private types)
- Removed ajv parser & dep (old stuff, only used in a single place)
- Fixed a bunch of missing types
- Removed as many unused types and schemas declarations as possible.
- Simplified the methods call in the dust client to all rely on the same doCall() method (exposed for custom use).

Next:
- Remove the exposition of the internal headers in `DustAPICredentials` and move this custom part as a wrapper in connectors.
- Do a review of methods that we should provide and ones that we should not in the client (for now, they are the same set as before).
- Push the npm package :)

Next next next / nice to have:
- Add linter to prevent importing private api stuff in v1
- Add dangers rule to warn if file is modified in sdks/js/types

## Risk

Type-checking is happy. I tested having conversations, uploading attachments etc..
However, the risk of breaking the public API remains.

## Deploy Plan

Deploy in front-edge, do some more testing, then deploy to prod and monitor.

### CODE REVIEW

File: front/lib/api/assistant/configuration.ts [hunk 1]
```diff
@@ -131,7 +131,13 @@ export async function searchAgentConfigurationsByName(
       },
     },
   });
-  return agentConfigurations || [];
```
Comment (Fraggle) [inline front/lib/api/assistant/configuration.ts:134]: we were sending the full model 🙄

File: front/lib/api/data_source_view.ts [hunk 2]
```diff
@@ -276,3 +279,51 @@ export async function getContentNodesForDataSourceView(
     total: contentNodesResult.total,
   });
 }
+
+export async function handlePatchDataSourceView(
```
Comment (Fraggle) [inline front/lib/api/data_source_view.ts:283]: Moved from a private endpoint to lib and refactored a bit to return `Result` instead of directly a next response.
Declare it's params type (here: `PatchDataSourceViewType`).

Then:
- in the public api endpoint, there is the public request schema (& response type)
- in the private api endpoint, there is the private request schema (& response type)
- both endpoints call this logic method.

This refactoring pattern has been applied a few times.

File: front/lib/api/data_sources.ts [hunk 3]
```diff
@@ -458,3 +466,211 @@ export async function upsertTable({
 
   return tableRes;
 }
+
+export async function handleDataSourceSearch({
```
Comment (Fraggle) [inline front/lib/api/data_sources.ts:470]: same as above.

File: front/lib/api/data_sources.ts [hunk 4]
```diff
@@ -458,3 +466,211 @@ export async function upsertTable({
...
+      },
+    });
+  }
+
+  return new Ok({
+    documents: data.value.documents,
+  });
+}
+
+export async function handleDataSourceTableCSVUpsert({
```
Comment (Fraggle) [inline front/lib/api/data_sources.ts:523]: same as above.

File: front/pages/api/poke/workspaces/[wId]/data_sources/[dsId]/search.ts [hunk 5]
```diff
@@ -52,7 +53,32 @@ async function handler(
...
+        return apiError(req, res, {
+          status_code: 400,
+          api_error: {
+            type: "invalid_request_error",
+            message: `Invalid request body: ${r.error.message}`,
+          },
+        });
+      }
+      const searchQuery = r.data;
+      const s = await handleDataSourceSearch({ searchQuery, dataSource });
```
Comment (Fraggle) [inline front/pages/api/poke/workspaces/[wId]/data_sources/[dsId]/search.ts:76]: see above to understand why the parsing of the request happens here now.

File: front/lib/api/assistant/configuration.ts [hunk 6]
```diff
@@ -131,7 +131,13 @@ export async function searchAgentConfigurationsByName(
...
+  const r = (
+    await Promise.all(
```
Comment (flvndvd) [inline front/lib/api/assistant/configuration.ts:135]: IIRC, you can directly call the `getAgentConfigurations` with all the ids to avoid the need for a promise.all wrapper here.

File: front/lib/api/assistant/configuration.ts [hunk 7]
```diff
@@ -131,7 +131,13 @@ export async function searchAgentConfigurationsByName(
...
+      agentConfigurations.map((c) => getLightAgentConfiguration(auth, c.sId))
+    )
+  ).filter((c) => c !== null) as LightAgentConfigurationType[];
```
Comment (flvndvd) [inline front/lib/api/assistant/configuration.ts:138]: We have a `removeNulls` helper.

File: front/lib/api/assistant/conversation.ts [hunk 8]
```diff
@@ -1953,9 +1953,9 @@ export function normalizeContentFragmentType({
 }: {
   contentType: SupportedContentFragmentType;
   url?: string;
-}) {
+}): SupportedContentFragmentType {
   // hack: for users creating content_fragments through our public API
-  if (contentType === "file_attachment") {
+  if ((contentType as string) === "file_attachment") {
```
Comment (flvndvd) [inline front/lib/api/assistant/conversation.ts:1958]: Why do you need the cast here?

File: front/lib/api/data_source_view.ts [hunk 9]
```diff
@@ -276,3 +279,51 @@ export async function getContentNodesForDataSourceView(
...
+  patchBody: PatchDataSourceViewType,
+  auth: Authenticator,
```
Comment (flvndvd) [inline front/lib/api/data_source_view.ts:285]: `nit`
Let's always keep `auth` as the first argument when present.

File: front/lib/api/data_source_view.ts [hunk 10]
```diff
@@ -276,3 +279,51 @@ export async function getContentNodesForDataSourceView(
...
+  dataSourceView: DataSourceViewResource
+): Promise<Result<DataSourceViewResource, APIErrorWithStatusCode>> {
+  if (!auth.isAdmin() && !auth.isBuilder()) {
+    throw new Err({
```
Comment (flvndvd) [inline front/lib/api/data_source_view.ts:289]: This seems very strange, throwing a custom `Err` type 🤔

File: front/lib/api/data_source_view.ts [hunk 11]
```diff
@@ -276,3 +279,51 @@ export async function getContentNodesForDataSourceView(
...
+    updateResultRes = await dataSourceView.updateParents(
+      parentsToAdd,
+      parentsToRemove
+    );
+  }
+
+  if (updateResultRes.isErr()) {
+    return new Err({
+      status_code: 500,
+      api_error: {
```
Comment (flvndvd) [inline front/lib/api/data_source_view.ts:318]: We are loosing the original error message here...

File: front/lib/api/data_sources.ts [hunk 12]
```diff
@@ -458,3 +466,211 @@ export async function upsertTable({
...
+        truncate,
+        useAppForHeaderDetection,
+      },
+    });
+    if (enqueueRes.isErr()) {
+      return new Err({
+        status_code: 500,
+        api_error: {
+          type: "data_source_error",
+          message:
```
Comment (flvndvd) [inline front/lib/api/data_sources.ts:603]: We are loosing the original error message here

File: front/package.json [hunk 13]
```diff
@@ -59,7 +59,6 @@
     "@tiptap/starter-kit": "^2.1.13",
     "@types/emoji-mart": "^3.0.10",
     "@uiw/react-textarea-code-editor": "^3.0.2",
-    "ajv": "^8.12.0",
```
Comment (flvndvd) [inline front/package.json:62]: 🔥

File: front/pages/api/poke/workspaces/[wId]/data_sources/[dsId]/search.ts [hunk 14]
```diff
@@ -52,7 +53,32 @@ async function handler(
...
+      if (s.isErr()) {
+        return apiError(req, res, s.error);
+      } else {
+        return res.json(s.value);
```
Comment (flvndvd) [inline front/pages/api/poke/workspaces/[wId]/data_sources/[dsId]/search.ts:80]: Don't you need to pass a status as well?

File: front/pages/api/poke/workspaces/[wId]/data_sources/[dsId]/search.ts [hunk 15]
```diff
@@ -52,7 +53,32 @@ async function handler(
 
   switch (req.method) {
     case "GET": {
-      return handleDataSourceSearch({ req, res, dataSource });
+      // I could not find a way to make the query params be an array if there is only one tag.
```
Comment (flvndvd) [inline front/pages/api/poke/workspaces/[wId]/data_sources/[dsId]/search.ts:56]: Why not using zod to handle this parsing?

File: front/pages/api/v1/w/[wId]/data_source_views/search.ts [hunk 16]
```diff
@@ -42,21 +29,19 @@ async function handler(
...
-        const pathError = reporter.formatValidationErrors(queryValidation.left);
+      const r = SearchDataSourceViewsRequestSchema.safeParse(req.query);
+
+      if (r.error) {
         return apiError(req, res, {
           status_code: 400,
           api_error: {
             type: "invalid_request_error",
-            message: `Invalid query parameters: ${pathError}`,
+            message: `Invalid request body: ${r.error.message}`,
```
Comment (flvndvd) [inline front/pages/api/v1/w/[wId]/data_source_views/search.ts:39]: It's actually the query 😉

File: front/pages/api/v1/w/[wId]/vaults/[vId]/data_source_views/[dsvId]/index.ts [hunk 17]
```diff
@@ -149,26 +146,56 @@ export type GetOrPostDataSourceViewsResponseBody = {
...
-  >,
+  res: NextApiResponse<WithAPIErrorResponse<DataSourceViewsResponseType>>,
   auth: Authenticator
 ): Promise<void> {
   const dataSourceView = await DataSourceViewResource.fetchById(
     auth,
     req.query.dsvId as string
   );
 
+  if (
```
Comment (flvndvd) [inline front/pages/api/v1/w/[wId]/vaults/[vId]/data_source_views/[dsvId]/index.ts:157]: Wait we were not checking this on the Public API 🤔?

File: front/pages/api/v1/w/[wId]/vaults/[vId]/data_source_views/[dsvId]/index.ts [hunk 18]
```diff
@@ -149,26 +146,56 @@ export type GetOrPostDataSourceViewsResponseBody = {
...
+        return apiError(req, res, {
+          status_code: 400,
+          api_error: {
+            type: "invalid_request_error",
+            message: `Invalid request body: ${r.error.message}`,
+          },
+        });
+      }
+
+      const r2 = await handlePatchDataSourceView(r.data, auth, dataSourceView);
```
Comment (flvndvd) [inline front/pages/api/v1/w/[wId]/vaults/[vId]/data_source_views/[dsvId]/index.ts:190]: r2 😱

File: front/pages/api/v1/w/[wId]/vaults/[vId]/data_source_views/[dsvId]/search.ts [hunk 19]
```diff
@@ -173,11 +174,35 @@ async function handler(
 
   switch (req.method) {
     case "GET": {
-      return handleDataSourceSearch({
-        req,
-        res,
+      // I could not find a way to make the query params be an array if there is only one tag.
```
Comment (flvndvd) [inline front/pages/api/v1/w/[wId]/vaults/[vId]/data_source_views/[dsvId]/search.ts:177]: Why not use zod?

File: front/pages/api/v1/w/[wId]/vaults/[vId]/data_sources/[dsId]/documents/[documentId]/index.ts [hunk 20]
```diff
@@ -372,24 +353,22 @@ async function handler(
...
             type: "invalid_request_error",
-            message: `Invalid request body: ${pathError}`,
+            message: `Invalid request body: ${r.error.message}`,
           },
         });
       }
 
       let sourceUrl: string | null = null;
-      if (bodyValidation.right.source_url) {
+      if (r.data.source_url) {
```
Comment (flvndvd) [inline front/pages/api/v1/w/[wId]/vaults/[vId]/data_sources/[dsId]/documents/[documentId]/index.ts:369]: `nit` We can destructure `data`.

File: sdks/js/src/index.ts [hunk 21]
```diff
@@ -354,6 +121,41 @@ export class DustAPI {
       : this._url;
   }
 
+  async doCall(args: DoCallArgsType) {
```
Comment (flvndvd) [inline sdks/js/src/index.ts:124]: `nit`
But if we plan to expose this method publicly I'd go with `request` to align on the most common libraries.

File: sdks/js/src/types.ts [hunk 22]
```diff
@@ -0,0 +1,1810 @@
+import moment from "moment-timezone";
```
Comment (flvndvd) [inline sdks/js/src/types.ts:1]: Not sure I'm a big fan of one big file, it's pretty hard to parse, and maintain. Can we maybe have one file per resource and use a "barrel file" to export them all?

**Review by flvndvd — COMMENTED** [review]
Left a first batch of comments.

File: front/pages/api/poke/workspaces/[wId]/data_sources/[dsId]/search.ts (hunk 15 above)
Comment (spolu) [reply front/pages/api/poke/workspaces/[wId]/data_sources/[dsId]/search.ts:56]: Let's not leak zod all over the place until we are commited to using it in front. I'm still very bearish on adding zod without an explicit commitment to remove iots, the risk of having a mixed bag of libraries to do the same thing is very high.

Comment (Fraggle) [reply front/pages/api/poke/workspaces/[wId]/data_sources/[dsId]/search.ts:56]: @flvndvd for the reason @spolu mentioned : i didn't want to start adding zod directly in front (or anywhere else).

Comment (Fraggle) [reply front/pages/api/poke/workspaces/[wId]/data_sources/[dsId]/search.ts:56]: @flvndvd but I'm happy to switch entierly from io-ts to zod in another PR if we want to.

File: front/pages/api/v1/w/[wId]/vaults/[vId]/data_source_views/[dsvId]/search.ts (hunk 19 above)
Comment (Fraggle) [reply front/pages/api/v1/w/[wId]/vaults/[vId]/data_source_views/[dsvId]/search.ts:177]: See above.

File: front/lib/api/assistant/configuration.ts [hunk 23]
```diff
@@ -118,7 +118,7 @@ export async function getAgentConfiguration(
 export async function searchAgentConfigurationsByName(
   auth: Authenticator,
   name: string
-): Promise<AgentConfiguration[] | []> {
+): Promise<LightAgentConfigurationType[]> {
```
[the remark this answers is not in the document]
Comment (Fraggle) [reply front/lib/api/assistant/configuration.ts:121]: We were sending the full agent model deserialized.

File: front/pages/api/v1/w/[wId]/vaults/[vId]/data_source_views/[dsvId]/index.ts (hunk 17 above)
Comment (spolu) [reply front/pages/api/v1/w/[wId]/vaults/[vId]/data_source_views/[dsvId]/index.ts:157]: Technically the dataSourceViewResource as REsourceWithVault should protect us here nonetheless right?

File: front/lib/api/assistant/configuration.ts (hunk 23 above)
[the remark this answers is not in the document]
Comment (spolu) [reply front/lib/api/assistant/configuration.ts:121]: ack

File: front/lib/api/assistant/configuration.ts (hunk 6 above)
Comment (Fraggle) [reply front/lib/api/assistant/configuration.ts:135]: Don't seems like it accepts a list of ids.

File: front/lib/api/assistant/conversation.ts (hunk 8 above)
Comment (Fraggle) [reply front/lib/api/assistant/conversation.ts:1959]: because SupportedContentFragmentType do not accept file_attachment

File: front/extension/app/src/pages/ConversationsPage.tsx [hunk 24]
```diff
@@ -44,29 +44,31 @@ export const ConversationsPage = ({
...
-      } else if (createdDate.isSameOrAfter(lastWeek)) {
-        groups["Last Week"].push(conversation);
-      } else if (createdDate.isSameOrAfter(lastMonth)) {
-        groups["Last Month"].push(conversation);
-      } else if (createdDate.isSameOrAfter(lastYear)) {
-        groups["Last 12 Months"].push(conversation);
-      } else {
-        groups["Older"].push(conversation);
+    conversations.forEach(
+      (conversation: ConversationWithoutContentPublicType) => {
```
Comment (flvndvd) [inline front/extension/app/src/pages/ConversationsPage.tsx:48]: `nit`
Type isn't required here.

File: front/lib/api/assistant/configuration.ts (hunk 6 above)
Comment (flvndvd) [reply front/lib/api/assistant/configuration.ts:136]: I meant for `getAgentConfigurations`  which is the method used under `getLightAgentConfiguration`:

https://github.com/dust-tt/dust/blob/fc26a61b098d0352a3b12222f945c87057b15c5c/front/lib/api/assistant/configuration.ts#L149-L159

File: front/lib/api/assistant/conversation.ts (hunk 8 above)
Comment (flvndvd) [reply front/lib/api/assistant/conversation.ts:1959]: Ah yeah legacy behavior.

File: front/lib/api/data_source_view.ts [hunk 25]
```diff
@@ -285,3 +287,48 @@ export async function getContentNodesForDataSourceView(
     total: contentNodesResult.total,
   });
 }
+
+export class NotAuthorizedError extends Error {}
```
Comment (flvndvd) [inline front/lib/api/data_source_view.ts:291]: Can we avoid creating one error class per type? We have an existing class `DustError` that partially covers those needs.

File: front/lib/api/data_sources.ts [hunk 26]
```diff
@@ -463,3 +470,195 @@ export async function upsertTable({
 
   return tableRes;
 }
+
+export class DataSourceError extends Error {}
```
Comment (flvndvd) [inline front/lib/api/data_sources.ts:474]: Same comment as above around Error.

File: front/pages/api/v1/w/[wId]/vaults/[vId]/data_source_views/[dsvId]/search.ts [hunk 27]
```diff
@@ -173,11 +177,51 @@ async function handler(
...
+          return apiError(req, res, {
+            status_code: 500,
+            api_error: {
+              type: "internal_server_error",
+              message: s.error,
+            },
+          });
+        }
+      } else {
+        return res.json(s.value);
```
Comment (flvndvd) [inline front/pages/api/v1/w/[wId]/vaults/[vId]/data_source_views/[dsvId]/search.ts:223]: Can we keep the existing patter of doing `res.status().json()`?

File: front/lib/api/data_source_view.ts (hunk 25 above)
Comment (Fraggle) [reply front/lib/api/data_source_view.ts:291]: Sure !

File: front/pages/api/v1/w/[wId]/vaults/[vId]/data_sources/[dsId]/search.ts [hunk 28]
```diff
@@ -194,7 +195,48 @@ async function handler(
...
+          return apiError(req, res, {
+            status_code: 500,
+            api_error: {
+              type: "internal_server_error",
+              message: s.error.message,
+            },
+          });
+        }
+      } else {
+        return res.json(s.value);
```
Comment (flvndvd) [inline front/pages/api/v1/w/[wId]/vaults/[vId]/data_sources/[dsId]/search.ts:238]: Still some left 😉

File: front/pages/api/w/[wId]/data_sources/[dsId]/tables/csv.ts [hunk 29]
```diff
@@ -91,181 +70,81 @@ async function handler(
...
+
+      const params = bodyValidation.right;
+
+      const r = await handleDataSourceTableCSVUpsert({
+        auth,
+        params,
+        dataSource,
+      });
+      if (r.isErr()) {
+        if (r.error.code === "missing_csv") {
```
Comment (flvndvd) [inline front/pages/api/w/[wId]/data_sources/[dsId]/tables/csv.ts:94]: `nit` switch/case? To ensure we update this part if we change the logic of `handleDataSourceTableCSVUpsert`

**Review by flvndvd — APPROVED** [review]
LGTM, left two last comments. Good to ship once addressed.

File: front/pages/api/v1/w/[wId]/vaults/[vId]/data_sources/[dsId]/search.ts (hunk 28 above)
Comment (Fraggle) [reply front/pages/api/v1/w/[wId]/vaults/[vId]/data_sources/[dsId]/search.ts:238]: 9 in the whole codebase. Let's do it in another PR.

Comment (flvndvd) [reply front/pages/api/v1/w/[wId]/vaults/[vId]/data_sources/[dsId]/search.ts:238]: Fine by me, as long as you track it 😉
